src/relay/backend/aot_executor_codegen.cc

```diff
@@ -656,6 +651,20 @@ class AOTExecutorCodegen : public MixedModeVisitor {
     auto storage_rewrite = tir::transform::StorageRewrite();
     mod_run = storage_rewrite(mod_run);
 
+    // The workspace for main function should be calculated after performing storage_rewrite for
+    // the top level TIR function.
+    auto workspace_byte_alignment =
+        target_host_->GetAttr<Integer>("workspace-byte-alignment").value_or(16);
+    Integer main_workspace_size = CalculateWorkspaceBytes(
```

Comment:
Does this imply the function metadata snaffled away by the callback from LowerTEPass above is also wrong? Perhaps we only need the entry for main anyway in which case we could get rid of UpdateFunctionMetadata entirely?

Reply:
Unfortunately `UpdateFunctionMetadata` creates the whole FunctionInfo object which includes `workspace_sizes`, `io_sizes`, `constant_sizes`, etc. I think we still need to do that during lowering...

Reply:
All of that function metadata is passed through and generated into the model library output, so it's still used - though I'm guessing it'd also be slightly wrong as `StorageRewrite` runs against the entire module. I'll take a look if I get a minute, this is more pressing as we use it to determine the size of the memory for the stack allocator.

Reply:
Hi @mbs-octoml @electriclilies ,

Ideally, we should run this when we have the the whole of IRModule in TIR which is a state currently only present in the AoT flow -- barring the ability to represent non-scalar constants in TIR (https://github.com/apache/tvm-rfcs/pull/22).

Once, all the executors could could start working based on TIR-only IRModule (after having done many progressive lowerings -- [a relevant discussion](https://github.com/apache/tvm-rfcs/pull/9#discussion_r688253505)), I think we could move UpdateFunctionMetadata (in the same time turning it to a pass) to the tail end of the pipeline.

In the mean time, we can still have the correct workspace sizes updated. For AoT, that should be the TIR main function just prior to the codegen and not the relay one used before.